content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,244 @@
+
+## Data clump
+One observation that we can make is that parameters are **a group of data that always comes together**. This data would make no sense when one piece of data is missing. We call it a set of **Data Clump**. For example, a start date and end date pair we would call a data clump or, as in this example, a data clump would be the first name and last name.
+
+In the "OOP" world, **objects contain data** and **classes contain behavior** (i.e. methods operating on data). It's cheap to create many small objects[^1] to manage and operate on the serial data. Additionally, we can define methods on the object and reuse them when necessary.
+[^1]: Many small objects could end up causing GC pressure. But high performance languages often have tricks up their sleeve like doingg [escape analysis](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/) to convert heap allocation to stack allocations or [expanding argument objects as primitive arguments](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/). These languages also generally have separate GC pathways for small objects and short-lived objects. In most cases, if you're just using an argument object locally in a high performance language, your runtime will manage to optimize most of the cost away.
+
+Let's try to create a new object to contain first name and last name.
+
+```rb
+class Author
+  attr_reader :first_name, :last_name
+ 
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+ 
+  def full_name
+    "#{last_name}, #{first_name}"
+  end
+end
+```
+The Author class also has two attributes, and the author object has one instance method call `full_name`. Now we can use the author object to replace the first name and last name.
+
+#### Before
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{year_published})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary("Bater", "Chen")
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+#### After
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{year_published})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary(Author.new("Bater", "Chen"))
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+The output of `summary` is the same, but the parameter reduces from two strings to one object. Well, you may doubt that it doesn't really matter, but imagine that you can follow the same pattern to reduce the parameter from 10 data into 1 object. That would be a huge impact.
+
+## Law of Demeter
+You may also have noticed that I relocated the logic of full name from book class to author class because a full name can perform independently without interaction with the book object. An **object should know less about each other** is said by the [Law of Demeter][2] or the "Least Knowledge Principle".
```

Comment:
I agree with the assessment that the law of demeter is an ill fit to explain this refactoring. LoD really boils down to enforced layering, and we're not changing the reach across layers.

Let's remove the reference to LoD and go with encapsulation.... or information hiding, but that doesn't fit our example really. Maybe just drop the justification-by-refering-to-authority, we give o full explanation of the why below. Then it could be (taking boaz' edits)

```suggestion
You may also have noticed that I relocated the logic of `full_name` from the `Book` class to the `Author`. This is because `full_name` can perform independently without interaction with the book object.
```